src/Session.cpp

```diff
@@ -117,69 +115,13 @@ void CSession::SetSupportedDecrypterURN(std::string& key_system)
   }
   m_KodiHost->SetLibraryPath(kodi::vfs::TranslateSpecialProtocol(specialpath).c_str());
 
-  std::array<std::string, 3> searchPaths =
-  {
-    kodi::vfs::TranslateSpecialProtocol("special://xbmcbinaddons/inputstream.adaptive/"),
-    kodi::vfs::TranslateSpecialProtocol("special://xbmcaltbinaddons/inputstream.adaptive/"),
-    kodi::addon::GetAddonInfo("path"),
-  };
-
-  std::vector<kodi::vfs::CDirEntry> items;
+  SSD::SSD_DECRYPTER* decrypter = CreateDecryptorInstance(m_KodiHost.get(), SSD::SSD_HOST::version);
+  const char* suppUrn(0);
 
-  for (auto searchPath : searchPaths)
+  if (decrypter && (suppUrn = decrypter->SelectKeySytem(m_kodiProps.m_licenseType.c_str())))
   {
-    LOG::Log(LOGDEBUG, "Searching for decrypters in: %s", searchPath.c_str());
-
-    if (!kodi::vfs::GetDirectory(searchPath, "", items))
-      continue;
-
-    for (auto item : items)
-    {
-      if (item.Label().compare(0, 4, "ssd_") && item.Label().compare(0, 7, "libssd_"))
-        continue;
-
-      bool success = false;
-      m_dllHelper = std::make_unique<kodi::tools::CDllHelper>();
-      if (m_dllHelper->LoadDll(item.Path()))
-      {
-#if defined(__linux__) && defined(__aarch64__) && !defined(ANDROID)
-        // On linux arm64, libwidevinecdm.so depends on two dynamic symbols:
-        //   __aarch64_ldadd4_acq_rel
-        //   __aarch64_swp4_acq_rel
-        // These are defined in libssd_wv.so, but to make them available in the main binary's PLT,
-        // we need RTLD_GLOBAL. LoadDll() above uses RTLD_LOCAL, so we use RTLD_NOLOAD here to
-        // switch the flags from LOCAL to GLOBAL.
```

Comment:
@lrusak did you follow any of this work earlier?

Going forward here will mean aarch64 widevine breaks, which is something [we knew would happen](https://github.com/xbmc/inputstream.adaptive/pull/1130#issuecomment-1428293770)

Do you think it's feasible we could get [these two symbols](https://github.com/xbmc/inputstream.adaptive/pull/1130/files#diff-cb49b4f5a6e2a38d51a696be771430404f8a6c7f7012b92d5d27da461825f790) patched into the kodi build itself?

Reply:
why can't we just have them in `inputsream.adaptive.so` ?

Reply:
when we `dlopen(libwidevinecdm.so)` from this addon, `ld` searches the main binary's PLT (linker table) to resolve the missing symbols. so it expects these two symbols to be there.

however when kodi `dlopen(inputstream.adaptive.so)` like it does for any addon, it uses `RT_LOCAL` (which is the default on linux anyway) so any symbols defined in this addon will stay confined to the add-on and not make their way to kodi's PLT

the trick i was using here was that `RT_NOLOAD | RT_GLOBAL` will force export entries up. we could continue to use that trick, but that would involve finding the path to `inputstream.adaptive.so` itself and calling dlopen on that now.

Reply:
Our choices are basically:

- do nothing and expect the user/OS to LD_PRELOAD some library that defines the syms
- define these syms inside kodi source/binary itself
- define inside isa.so and use dlopen reload hack
- ship another .so with isa that contains only these symbols, that we then load with dlopen(RT_GLOBAL)

Reply:
is there an issue with using `RT_NOLOAD | RT_GLOBAL` when loading the add-on shared library from Kodi?

Reply:
using `RT_GLOBAL` for an add-on system would not be ideal, because it could mean functions defined in different addons could conflict and interact with each other.

on macOS `RT_GLOBAL` is the default, which is something [I had to change](https://github.com/xbmc/xbmc/pull/20219) because it affected how different versions of an add-on written in golang would conflict with each other when upgrading at runtime

Reply:
For the time being LD_PRELOAD seems like a simple solution. Maybe we can document a simple source code example, compilation instructions and usage?

Reply:
as said if you remove it you are breaking widevine on all Kodi OS (LE/CE etc...), RPI OS and Linux distributions,
reading on forum threads the use of LD_PRELOAD is not a so easy solution for all OS
because they need to build a selfmade library to do a similar thing etc...

it is also complicated for users who do not know what this is to have to teach them to install a kind of preloadon their OS's to make Kodi works with widevine...

doing such a thing you must first discuss on slack at least with the LibreElec and CoreElec team, and also get more informations for RPI OS and in general for Linux OS

I don't know if this cleanup is worth doing it
given the expectations of removing this ssd library

the objective is to move all ssd code content to main project,
and keep this shared library only to contains the widevine aarch64 fix

where here i have attached a example scheme
https://github.com/xbmc/inputstream.adaptive/issues/736#issuecomment-1445342957